File: python/taflow/metrics/net_profit.py
"""Net raw profit and loss metric."""
from __future__ import annotations
from typing import Any
from .._native.metrics import NetProfit as _Native
from ._input import as_metric_series


class NetProfit:
    """Sum raw P&L observations with their original signs.

    The independent oracle is NumPy ``sum`` and the result equals Gross Profit
    plus signed Gross Loss. Warm-up requires one usable observation; a nonempty
    all-breakeven sample returns zero. Only raw non-cumulative period P&L and
    realized closed-trade P&L are accepted. No capital conversion or
    annualization occurs. Rust owns compensated O(1) streaming state.
    """
    def __init__(self, nan_policy: str = "omit") -> None:
        """Initialize an empty configured metric."""
        self._state = _Native(nan_policy)

    def from_pnl(self, pnl: Any, *, column: str | None = None) -> "NetProfit":
        """Append chronological pnl observations and return this metric."""
        self._state.from_pnl(as_metric_series(pnl, column=column))
        return self

    def from_trades(self, trades: Any, *, column: str | None = None) -> "NetProfit":
        """Append chronological trades observations and return this metric."""
        self._state.from_trades(as_metric_series(trades, column=column))
        return self

    def append(self,value:float)->"NetProfit":
        """Append one P&L observation and return this metric."""
        self._state.append(float(value));return self
    def extend(self,values:Any,*,column:str|None=None)->"NetProfit":
        """Append P&L observations and return this metric."""
        self._state.extend(as_metric_series(values,column=column));return self
    @property
    def value(self)->float|None:
        """Return net profit, or ``None`` during warm-up."""
        return self._state.value
    def compute(self)->float|None:
        """Return current scalar without replaying input."""
        return self._state.compute()
    def reset(self)->"NetProfit":
        """Clear state, preserve the P&L domain, and return this metric."""
        self._state.reset();return self
    def __len__(self)->int:
        """Return valid P&L observation count delegated to Rust."""
        return len(self._state)


__all__=["NetProfit"]

File: crates/taflow-metrics/src/metrics/net_profit.rs
use crate::{MetricError, MetricInputKind, MetricInputState, MetricResult, NanPolicy};

/// Net sum of raw period or closed-trade profit and loss observations.
#[derive(Debug, Clone)]
pub struct NetProfit {
    input: MetricInputState,
    sum: f64,
    compensation: f64,
}
impl NetProfit {
    /// Construct an empty raw P&L or closed-trade state.
    pub fn new(nan_policy: NanPolicy) -> MetricResult<Self> {
        Ok(Self {
            input: MetricInputState::unbound(nan_policy),
            sum: 0.0,
            compensation: 0.0,
        })
    }
    /// Append one P&L observation and return net profit.
    pub fn append(&mut self, value: f64) -> MetricResult<Option<f64>> {
        if let Some(observation) = self.input.append(value)? {
            let adjusted = observation - self.compensation;
            let next = self.sum + adjusted;
            self.compensation = (next - self.sum) - adjusted;
            self.sum = next;
        }
        Ok(self.value())
    }
    /// Append observations through the same persistent state.
    pub fn extend(&mut self, values: &[f64]) -> MetricResult<Option<f64>> {
        self.input.extend(values, |observation| {
            let adjusted = observation - self.compensation;
            let next = self.sum + adjusted;
            self.compensation = (next - self.sum) - adjusted;
            self.sum = next;
            Ok(())
        })?;
        Ok(self.value())
    }
    /// Return gross profit plus signed gross loss.
    pub fn value(&self) -> Option<f64> {
        (!self.input.is_empty()).then_some(self.sum)
    }
    /// Return current scalar without replay.
    pub fn compute(&self) -> Option<f64> {
        self.value()
    }
    /// Reset while preserving input domain.
    pub fn reset(&mut self) {
        self.input.reset();
        self.sum = 0.0;
        self.compensation = 0.0;
    }
    /// Return valid observation count.
    pub fn len(&self) -> usize {
        self.input.len()
    }
    /// Return whether no observations were processed.
    pub fn is_empty(&self) -> bool {
        self.input.is_empty()
    }
}

crate::impl_pnl_trade_metric_lifecycle!(NetProfit);
